File: Medium/2026-07-09-3532-Path-Existence-Queries-in-a-Graph-I/solution.py
class Solution:
    def distanceLimitedPathsExist(self, n: int, nums: List[int], maxDiff: int, queries: List[List[int]]) -> List[bool]:
        parent = list(range(n))

        def find(x):
            if parent[x] != x:
                parent[x] = find(parent[x])
            return parent[x]

        def union(x, y):
            rootX = find(x)
            rootY = find(y)
            if rootX != rootY:
                parent[rootY] = rootX

        edges = []
        for i in range(n):
            for j in range(i + 1, n):
                if abs(nums[i] - nums[j]) <= maxDiff:
                    edges.append((i, j))

        for u, v in edges:
            union(u, v)

        result = []
        for u, v in queries:
            result.append(find(u) == find(v))

        return result

File: Medium/2026-07-09-3532-Path-Existence-Queries-in-a-Graph-I/solution.java
import java.util.*;

class Solution {
    public boolean[] distanceLimitedPathsExist(int n, int[] nums, int maxDiff, int[][] queries) {
        List<int[]> edges = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j < n && nums[j] - nums[i] <= maxDiff; j++) {
                edges.add(new int[]{i, j});
            }
        }

        UnionFind uf = new UnionFind(n);
        for (int[] edge : edges) {
            uf.union(edge[0], edge[1]);
        }

        boolean[] result = new boolean[queries.length];
        for (int i = 0; i < queries.length; i++) {
            result[i] = uf.find(queries[i][0]) == uf.find(queries[i][1]);
        }

        return result;
    }

    private static class UnionFind {
        private int[] parent;
        private int[] rank;

        public UnionFind(int size) {
            parent = new int[size];
            rank = new int[size];
            for (int i = 0; i < size; i++) {
                parent[i] = i;
                rank[i] = 1;
            }
        }

        public int find(int x) {
            if (parent[x] != x) {
                parent[x] = find(parent[x]);
            }
            return parent[x];
        }

        public void union(int x, int y) {
            int rootX = find(x);
            int rootY = find(y);
            if (rootX != rootY) {
                if (rank[rootX] > rank[rootY]) {
                    parent[rootY] = rootX;
                } else if (rank[rootX] < rank[rootY]) {
                    parent[rootX] = rootY;
                } else {
                    parent[rootY] = rootX;
                    rank[rootX]++;
                }
            }
        }
    }
}